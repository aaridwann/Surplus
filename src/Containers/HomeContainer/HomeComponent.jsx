import React, { useMemo, useRef, useState } from 'react';
import {
	Animated,
	Image,
	SafeAreaView,
	ScrollView,
	Text,
	TouchableOpacity,
	View,
} from 'react-native';
import SearchComponent from '../../Components/SearchComponent/SearchComponent';
import Constants from '../../Utils/Constants/Constants';
import { scaleSize } from '../../Utils/Scale/Scale';
import HighlightComponent from '../Highlight/HighlightComponent';
import { get } from 'lodash';
import FeedsComponent from '../Feeds/Feeds';
import DropDownContainer from '../../Components/FilterComponent/FilterComponent';
import { Ionicons } from '@expo/vector-icons';
import { SimpleLineIcons } from '@expo/vector-icons';
import * as Animatable from 'react-native-animatable';
import Styles from './HomeComponent.Style';

const _renderBackground = ({ username }, { Logout }) => {
	const [showOption, setShowOption] = useState(false);
	const { IMG } = Constants;

	return (
		<Animatable.View
			animation={'slideInDown'}
			duration={1500}
			style={Styles.background.container}
		>
			<View style={Styles.background.leftContentWrapper}>
				<SimpleLineIcons
					style={Styles.background.icon}
					name="diamond"
					size={scaleSize(18)}
					color="white"
				/>
				<Text style={Styles.background.username}>{username}</Text>
			</View>
			<Image style={Styles.background.img} source={IMG.SURPLUS_LOGO} />
			<View style={Styles.background.optionWrapper}>
				{showOption && (
					<TouchableOpacity onPress={Logout}>
						<Text style={Styles.background.optionsText}>Logout</Text>
					</TouchableOpacity>
				)}
				<TouchableOpacity onPress={() => setShowOption((prev) => !prev)}>
					<Ionicons
						style={Styles.background.ellipsisIcon}
						name="ellipsis-vertical"
						size={scaleSize(24)}
						color={'white'}
					/>
				</TouchableOpacity>
			</View>
		</Animatable.View>
	);
};
const _renderHightLightComponent = (dataHighlight) => (
	<View style={Styles.highlightWrapper}>
		<HighlightComponent data={dataHighlight} />
	</View>
);
const _renderFeedsWrapper = (dataFeeds, go) => (
	<Animatable.View
		duration={1000}
		// easing={'ease-in'}
		// animation={'slideInUp'}
		style={Styles.feedsWrapper}
	>
		<FeedsComponent data={dataFeeds} onPress={go} />
	</Animatable.View>
);
const _renderSearchWrapper = (animatedProps, search) => (
	<Animated.View style={animatedProps}>
		<Animatable.View duration={1500} animation={'slideInRight'}>
			<SearchComponent onChange={(val) => search(val)} />
		</Animatable.View>
	</Animated.View>
);

/**
 *
 * @param {State, method} Props
 * @returns Parent Component Home
 */
const HomeComponent = (Props) => {
	const { state, method } = Props;
	const [screen, setScreen] = useState(0);
	const [show, setShow] = useState(false);
	const dataHighlight = useMemo(() => get(state, 'data', []), [state]);
	const dataFeeds = useMemo(() => get(state, 'dataFeeds', []), [
		state.dataFeeds,
	]);
	const animatedTopValue = useRef(
		new Animated.Value(show || screen === 0 ? 80 : -80)
	)?.current;
	const handleScroll = (event) => {
		const { contentOffset, layoutMeasurement, contentSize } = event.nativeEvent;
		const y = get(contentOffset, 'y', 0);

		if (y > screen) {
			setShow(false);
			setScreen(y);
		} else {
			setShow(true);
			setScreen(() => y);
		}

		const isScrolledToBottom =
			contentOffset.y + layoutMeasurement.height >= contentSize.height - 10;

		if (isScrolledToBottom) method.LoadMore();
	};
	const animatedProps = [
		Styles.searchContainer,
		{ transform: [{ translateY: animatedTopValue }] },
	];
	const scrollProps = {
		onScroll: handleScroll,
		contentContainerStyle: Styles.scrollViewContainer,
		showsVerticalScrollIndicator: false,
	};
	const filterComponentProps = {
		selectCountry: (e) => method?.SelectFilter({ type: 'country', value: e }),
		selectGender: (e) => method?.SelectFilter({ type: 'gender', value: e }),
	};

	React.useEffect(() => {
		Animated.timing(animatedTopValue, {
			toValue: show || screen === 0 ? 80 : -180,
			duration: 80,
			useNativeDriver: false,
		})?.start();
	}, [show, screen, animatedTopValue]);

	return (
		<SafeAreaView style={Styles.container}>
			{_renderSearchWrapper(animatedProps, method.Search)}
			<ScrollView {...scrollProps}>
				{_renderBackground(state, method)}
				{_renderHightLightComponent(dataHighlight)}
				{DropDownContainer({ ...filterComponentProps })}
				{_renderFeedsWrapper(dataFeeds, method.GoToDetails)}
			</ScrollView>
		</SafeAreaView>
	);
};

export default HomeComponent;
