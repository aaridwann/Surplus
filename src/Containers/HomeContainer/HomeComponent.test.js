import { render } from '@testing-library/react-native';
import React from 'react';
import HomeComponent from './HomeComponent';

describe('HomeComponent', () => {
	beforeAll(() => {
		jest.mock(
			'../../Components/FilterComponent/FilterComponent',
			() => 'Filter'
		);
	});

	const MockData = [
		{
			picture: {
				medium: 'https://example.com/image1.jpg',
				large: 'https://example.com/image1.jpg',
			},
			name: {
				first: 'John',
			},
		},
		{
			picture: {
				medium: 'https://example.com/image2.jpg',
				large: 'https://example.com/image2.jpg',
			},
			name: {
				first: 'Jane',
			},
		},
	];
	const Props = {
		state: {
			loading: false,
			data: MockData,
			dataFeeds: MockData,
			dataFeedsBySearch: [],
			pageFeeds: 3,
			country: '',
			gender: '',
		},
		method: {
			LoadMore: jest.fn(),
			Search: jest.fn(),
			SelectFilter: jest.fn(),
			GoToDetails: jest.fn(),
			Logout: jest.fn(),
		},
	};

	it('Should snapshot correct layout', () => {
		expect(render(<HomeComponent {...Props} />)).toMatchSnapshot();
	});
});
