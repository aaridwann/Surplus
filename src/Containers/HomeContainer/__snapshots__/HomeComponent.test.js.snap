// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeComponent Should snapshot correct layout 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "white",
      "color": "gray",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "width": "100%",
      }
    }
    onScroll={[Function]}
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "marginTop": 70,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "minHeight": 160,
              "padding": 16,
              "paddingVertical": 20,
              "width": "100%",
            }
          }
          testID="highlight-component"
        >
          <Text
            style={
              Object {
                "color": "#50b8a8",
                "fontSize": 30,
                "fontWeight": "700",
              }
            }
          >
            Highlight
          </Text>
          <RCTScrollView
            contentContainerStyle={
              Object {
                "marginTop": 10,
              }
            }
            data={
              Array [
                Object {
                  "name": Object {
                    "first": "John",
                  },
                  "picture": Object {
                    "large": "https://example.com/image1.jpg",
                    "medium": "https://example.com/image1.jpg",
                  },
                },
                Object {
                  "name": Object {
                    "first": "Jane",
                  },
                  "picture": Object {
                    "large": "https://example.com/image2.jpg",
                    "medium": "https://example.com/image2.jpg",
                  },
                },
              ]
            }
            getItem={[Function]}
            getItemCount={[Function]}
            horizontal={true}
            keyExtractor={[Function]}
            onContentSizeChange={[Function]}
            onLayout={[Function]}
            onMomentumScrollBegin={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            removeClippedSubviews={false}
            renderItem={[Function]}
            scrollEventThrottle={50}
            showsHorizontalScrollIndicator={false}
            stickyHeaderIndices={Array []}
            viewabilityConfigCallbackPairs={Array []}
          >
            <View>
              <View
                onLayout={[Function]}
                style={
                  Array [
                    Object {
                      "flexDirection": "row",
                    },
                    null,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "color": "gray",
                      "marginHorizontal": 8,
                    }
                  }
                  testID="highlight-item-0"
                >
                  <Image
                    source={
                      Object {
                        "uri": "https://example.com/image1.jpg",
                      }
                    }
                    style={
                      Object {
                        "backgroundColor": "#50b8a8",
                        "borderColor": "#50b8a8",
                        "borderRadius": 50,
                        "borderWidth": 2.5,
                        "height": 100,
                        "width": 100,
                      }
                    }
                    testID="highlight-image-0"
                  />
                  <Text
                    style={
                      Object {
                        "color": "#50b8a8",
                        "fontSize": 30,
                        "fontWeight": "500",
                        "marginVertical": 4,
                      }
                    }
                    testID="highlight-text-0"
                  >
                    John
                  </Text>
                </View>
              </View>
              <View
                onLayout={[Function]}
                style={
                  Array [
                    Object {
                      "flexDirection": "row",
                    },
                    null,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "color": "gray",
                      "marginHorizontal": 8,
                    }
                  }
                  testID="highlight-item-1"
                >
                  <Image
                    source={
                      Object {
                        "uri": "https://example.com/image2.jpg",
                      }
                    }
                    style={
                      Object {
                        "backgroundColor": "#50b8a8",
                        "borderColor": "#50b8a8",
                        "borderRadius": 50,
                        "borderWidth": 2.5,
                        "height": 100,
                        "width": 100,
                      }
                    }
                    testID="highlight-image-1"
                  />
                  <Text
                    style={
                      Object {
                        "color": "#50b8a8",
                        "fontSize": 30,
                        "fontWeight": "500",
                        "marginVertical": 4,
                      }
                    }
                    testID="highlight-text-1"
                  >
                    Jane
                  </Text>
                </View>
              </View>
            </View>
          </RCTScrollView>
        </View>
      </View>
      <View
        collapsable={false}
        forwardedRef={[Function]}
        style={
          Object {
            "marginTop": 20,
            "transform": Array [
              Object {
                "translateY": 100,
              },
            ],
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "center",
            }
          }
          testID="feeds-component"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#50b8a8",
                "height": 300,
                "justifyContent": "center",
                "margin": 5,
                "overflow": "hidden",
                "width": 340,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://example.com/image1.jpg",
                }
              }
              style={
                Object {
                  "height": "100%",
                  "margin": 4,
                  "width": "100%",
                }
              }
            />
            <Text
              onPress={[Function]}
              style={
                Object {
                  "alignSelf": "flex-start",
                  "bottom": 0,
                  "color": "white",
                  "fontSize": 60,
                  "justifyContent": "flex-start",
                  "margin": 4,
                  "position": "absolute",
                  "textTransform": "capitalize",
                }
              }
              testID="feed-name"
            >
              John
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#50b8a8",
                "height": 300,
                "justifyContent": "center",
                "margin": 5,
                "overflow": "hidden",
                "width": 340,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://example.com/image2.jpg",
                }
              }
              style={
                Object {
                  "height": "100%",
                  "margin": 4,
                  "width": "100%",
                }
              }
            />
            <Text
              onPress={[Function]}
              style={
                Object {
                  "alignSelf": "flex-start",
                  "bottom": 0,
                  "color": "white",
                  "fontSize": 60,
                  "justifyContent": "flex-start",
                  "margin": 4,
                  "position": "absolute",
                  "textTransform": "capitalize",
                }
              }
              testID="feed-name"
            >
              Jane
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
